clang/unittests/ASTMatchers/ASTMatchersNarrowingTest.cpp

```diff
@@ -2503,6 +2503,28 @@ TEST_P(ASTMatchersTest, IsDelegatingConstructor) {
       cxxConstructorDecl(isDelegatingConstructor(), parameterCountIs(1))));
 }
 
+TEST_P(ASTMatchersTest, MatchesString) {
+  StatementMatcher Literal = stringLiteral(matchesString("foo.*"));
+  EXPECT_TRUE(matches("const char* a = \"foo\";", Literal));
+  EXPECT_TRUE(matches("const char* b = \"foobar\";", Literal));
+  EXPECT_TRUE(matches("const char* b = \"fo\"\"obar\";", Literal));
+  EXPECT_TRUE(notMatches("const char* c = \"bar\";", Literal));
+  // test embedded nulls
+  StatementMatcher Literal2 = stringLiteral(matchesString("bar"));
+  EXPECT_TRUE(matches("const char* b = \"foo\\0bar\";", Literal2));
+  EXPECT_TRUE(notMatches("const char* b = \"foo\\0b\\0ar\";", Literal2));
+}
+
+TEST(MatchesString, MatchesStringPrefixed) {
+  StatementMatcher Literal = stringLiteral(matchesString("foo.*"));
+  EXPECT_TRUE(matchesConditionally("const char16_t* a = u\"foo\";", Literal,
+                                   true, {"-std=c++11"}));
+  EXPECT_TRUE(matchesConditionally("const char32_t* a = U\"foo\";", Literal,
+                                   true, {"-std=c++11"}));
+  EXPECT_TRUE(matchesConditionally("const wchar_t* a = L\"foo\";", Literal,
+                                   true, {"-std=c++11"}));
```

Comment:
CC @cor3ntin @tahonermann 

Should these actually match? I'm not certain how we should handle these in general given that the regex is a string literal with no encoding information and it's matching against source string literals which do have a runtime encoding. Because this is AST matching, I suppose the idea is that the regex is UTF-8 and the source is treated as UTF-8, so the encoding prefixes shouldn't matter. Do I have that right?

Reply:
The matcher seems to be using the value produced by `StringLiteral::outputString` which will "work" - until we support non unicode encoding - but `StringLiteral::outputString` is really designed for debug, so wchar etc are going to be escaped. I think this is going to be a difficult to use interface.

We would be better off adding a way to convert back the string literal to non-escaped UTF-8 ... except numeric escape sequence would not be preserved (and might lead to replacement characters)

Reply:
Now I use getBytes instead https://github.com/llvm/llvm-project/pull/102152/commits/22ad5bff393984cfdf536134dbdb8e5cebdd8ddb
I think you're right; we shouldn't match strings with different encoding prefixes. Do you have any suggestions? For instance, should we write a separate matcher for each encoding format of string literals?

Reply:
Sorry, this PR dropped off my radar entirely!

I think that for right now, we should maybe prohibit matching so that we allow `char` and `char8_t` but prohibit `char16_t` and `char32_t` matching. But @cor3ntin brings up a good point about numeric escape sequences. We should probably have test coverage for something like matching `"foo"` against `"\x66\x6f\x6f"` to ensure we match. Or do we expect that to not match because we want `matchesString()` to match on the syntactic form of the characters in the source rather than the semantic form of the characters?